File: 201_300/264_Ugly_Number_II/Solution.cpp
class Solution {
public:
    int nthUglyNumber(int n) {
        if (n < 1) return 0;
        int* uglyNumbers = new int[n];
        uglyNumbers[0] = 1;
        int next = 1;
        
        int *p2 = uglyNumbers;
        int *p3 = uglyNumbers;
        int *p5 = uglyNumbers;
        
        for (;next < n;next++) {
            int min1 = *p2 * 2 < *p3 * 3 ? *p2 * 2 : *p3 * 3;
            int minUgly = min1 < *p5 * 5 ? min1 : *p5 * 5; 
            uglyNumbers[next] = minUgly;
            
            if (*p2*2 <= uglyNumbers[next]) *p2++;
            if (*p3*3 <= uglyNumbers[next]) *p3++;
            if (*p5*5 <= uglyNumbers[next]) *p5++;
    
        }
        
        return uglyNumbers[next - 1];
    }
    
};

File: 201_300/264_Ugly_Number_II/Solution.rs
impl Solution {
    pub fn nth_ugly_number(n: i32) -> i32 {
        if n < 1 {
            return 0;
        }
        let mut dp = vec![1_i32; n as usize];
        let (mut p2, mut p3, mut p5) = (0, 0, 0);
        for next in 1..n as usize {
            dp[next] = (dp[p2] * 2).min(dp[p3] * 3).min(dp[p5] * 5);
            if dp[next] == dp[p2] * 2 {
                p2 += 1;
            }
            if dp[next] == dp[p3] * 3 {
                p3 += 1;
            }
            if dp[next] == dp[p5] * 5 {
                p5 += 1;
            }
        }
        dp[n as usize - 1]
    }
}
